CMakeLists.txt

```diff
@@ -2,85 +2,53 @@ cmake_minimum_required(VERSION 3.15)
 
 # Set the project name to your project name, my project isn't very descriptive
 project(myproject CXX)
-include(cmake/StandardProjectSettings.cmake)
-include(cmake/PreventInSourceBuilds.cmake)
 
-# Link this 'library' to set the c++ standard / compile-time options requested
-add_library(project_options INTERFACE)
+# Add cmakelib
+include(FetchContent)
+FetchContent_Declare(cmakelib URL https://github.com/aminya/cmakelib/archive/refs/heads/main.zip)
+FetchContent_MakeAvailable(cmakelib)
+include(${cmakelib_SOURCE_DIR}/Index.cmake)
+
+# Initialize cmakelib
+# uncomment the options to enable them
+cmakelib(
+    ENABLE_CACHE
+    ENABLE_CONAN
+    WARNINGS_AS_ERRORS
+    # ENABLE_CPPCHECK
+    # ENABLE_CLANG_TIDY
+    # ENABLE_INCLUDE_WHAT_YOU_USE
+    # ENABLE_COVERAGE
+    # ENABLE_PCH
+    # ENABLE_DOXYGEN
+    # ENABLE_IPO
+    # ENABLE_USER_LINKER
+    # ENABLE_BUILD_WITH_TIME_TRACE
+    # ENABLE_UNITY
+    # ENABLE_SANITIZER_ADDRESS
+    # ENABLE_SANITIZER_LEAK
+    # ENABLE_SANITIZER_UNDEFINED_BEHAVIOR
+    # ENABLE_SANITIZER_THREAD
+    # ENABLE_SANITIZER_MEMORY
```

Comment:
I can see what you're doing here, and I think there's some value in changing some of these options into constants that cannot be changed without modifying CMake code.

However, I find it very useful to be able to turn many of these options on and off at will, without modifying any code. For instance, if you're using the sanitizers, the only sensible way to use them is by turning them on and off. Most (all?) of them are incompatible with one another, and if two are turned on at once, they will interfere with each other. If you're using them, they need to be turned on and off one at a time. It is true that you can just uncomment the options you need, whenever you need them, but for a lot of use cases, I think it's more practical to do it with a CMake option.

If you need a CI build that runs the sanitizers, how would you turn them on without affecting the other CI builds? How about a build specifically to run the static analyzers, or doxygen to build your documentation? If you need all of these features, are you just supposed to turn them all on at once? Some of these features make more sense to run in CI (ie doxygen, coverage) than locally on a developer's machine. It is not appropriate for me to have `ENABLE_COVERAGE` turned on while I am trying to debug some C++ code, and I don't like having to reach in and modify CMake files so I can get some work done.

I think it might make sense to have `ENABLE_CACHE`, `WARNINGS_AS_ERRORS`, and maybe `ENABLE_UNITY` to be hardcoded, but I can think of reasons for most of the others to remain as CMake configuration options. I am not sure about `ENABLE_PCH`, `ENABLE_UNITY` and `ENABLE_IPO`; this is likely due to my unfamiliarity with those tools and/or lack of imagination.

Reply:
The benefit of using function arguments is that it allows building anything you want on top of it. For example, in case anyone wants to use global mutable states for managing their build, it is possible to add an option on top of the function arguemtns. 

Let's say for sanitizers:
```cmake
option(ENABLE_ADDRESS_SANITIZER_OPT "enable address sanitizer" OFF)

if(ENABLE_ADDRESS_SANITIZER_OPT)
	set(ENABLE_ADDRESS_SANITIZER ON)
else()
	set(ENABLE_ADDRESS_SANITIZER OFF)
endif()

cmakelib(
	# ...
	ENABLE_ADDRESS_SANITIZER 
)
```

Reply:
I like the use of function arguments here, that's not what I'm objecting to. For me, much of the value of this project is that the boilerplate code is already written, and I don't have to go in and add all those cmake options myself. After this change, I will have very little reason to use this template at all; I might as well just use cmakelib instead and copy paste the boilerplate in the readme.

If this project is supposed to be a quick-start to having a C++ project, I think this change is moving directly against that goal. Maybe I'm misunderstanding something, maybe I'm being unnecessarily stubborn; in either case, I think we need input from @lefticus to resolve this

Reply:
I don't understand why you think this is against the goal. Rust has 68000 packages that are all built using a single command `cargo build`. The reason is that everything is declarative by default. In contrast C++ only has 1000 packages on Conan.

Why do you think it is good to advertise using global mutable states and boilerplate code? I argue that declarative build system contributes to the reproducibility of the builds making it easier to write good packages. 

As I said, if anyone wants to do exotic things, they can do it. But I don't think we should advertise that. Also, Cmakelib is available for downloading. People can still download and modify that as they wish.

Reply:
```cmake
option(ENABLE_ADDRESS_SANITIZER_OPT "enable address sanitizer" OFF)

if(ENABLE_ADDRESS_SANITIZER_OPT)
	set(ENABLE_ADDRESS_SANITIZER ON)
else()
	set(ENABLE_ADDRESS_SANITIZER OFF)
endif()

cmakelib(
	# ...
	ENABLE_ADDRESS_SANITIZER 
)
```
This can NOT work!

Reply:
If you really want global options, we can add a file that can be optionally included to make those function arguments available as global options.

Here is a simple code that does the trick:
https://github.com/aminya/cmakelib/pull/9

Reply:
> I don't understand why you think this is against the goal.

One of the stated goals of this project is to promote the use of sanitizers, coverage, static analyzers, etc. This is related to ["Use the tools available"](https://github.com/cpp-best-practices/cppbestpractices/blob/master/02-Use_the_Tools_Available.md). This change makes the tools unreachable from the command line, and requires code changes to turn the tools on and off.

Please keep in mind that a lot of people use this project (it has 1.4k stars at the moment), and depend on these features. They will not be pleased when you break them.

> I argue that declarative build system contributes to the reproducibility of the builds making it easier to write good packages.

This is a good point, and I agree with it. Using global options means that there will be variable name conflicts when one cpp_starter_project includes another cpp_starter_project. This is problem worth solving, and I don't have a ready solution for it.

I think it is really unproductive to continue to debate this without @lefticus' input. Even if I were to approve this PR, you will need his approval to merge, and it's hard for me to imagine that he would want to remove these features.

Reply:
I just showed you a way to make things optionally global. Something like this

```cmake
# Make things global
include(${cmakelib_SOURCE_DIR}/GlobalOptions.cmake)

cmakelib(
  ${ENABLE_CACHE}
)
```

Reply:
Fixed in https://github.com/aminya/cmakelib/pull/9